game-core/src/main/java/games/strategy/engine/posted/game/pbf/NodeBbTokenGenerator.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.message.BasicNameValuePair;
+import org.apache.http.util.EntityUtils;
+import org.snakeyaml.engine.v2.api.Load;
+import org.snakeyaml.engine.v2.api.LoadSettings;
+
+@AllArgsConstructor
+public class NodeBbTokenGenerator {
+  private final String forumUrl;
+  private final Load load = new Load(LoadSettings.builder().build());
+
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  @Getter
+  public static class TokenInfo {
+    @Nonnull private final String token;
+    private final int userId;
+  }
+
+  public TokenInfo generateToken(
+      final String username, final String password, @Nullable final String otp) {
+    Preconditions.checkNotNull(username);
+    Preconditions.checkNotNull(password);
+
+    try (CloseableHttpClient client = HttpClients.custom().disableCookieManagement().build()) {
+      final int userId = getUserId(client, username);
+      return new TokenInfo(getToken(client, userId, password, otp), userId);
+    } catch (final IOException e) {
+      throw new RuntimeException("Failed to retrieve login token", e);
+    }
+  }
+
+  public void revokeToken(final String token, final int userId) {
+    try (CloseableHttpClient client = HttpClients.custom().disableCookieManagement().build()) {
+      deleteToken(client, userId, token);
+    } catch (final IOException e) {
+      throw new RuntimeException("Failed to revoke login token", e);
+    }
+  }
+
+  private void deleteToken(final CloseableHttpClient client, final int userId, final String token)
+      throws IOException {
+    final HttpDelete httpDelete =
+        new HttpDelete(forumUrl + "/api/v2/users/" + userId + "/tokens/" + token);
+    HttpProxy.addProxy(httpDelete);
+    httpDelete.addHeader("Authorization", "Bearer " + token);
+    client.execute(httpDelete).close(); // ignore errors, execute and then close
+  }
+
+  private int getUserId(final CloseableHttpClient client, final String username)
+      throws IOException {
+    final Map<?, ?> jsonObject = queryUserInfo(client, username);
+    checkUser(jsonObject, username);
+    return (Integer) jsonObject.get("uid");
+  }
+
+  private void checkUser(final Map<?, ?> jsonObject, final String username) {
+    if (!jsonObject.containsKey("uid")) {
+      throw new IllegalStateException(String.format("User %s doesn't exist.", username));
+    }
+    if (1 == (Integer) jsonObject.get("banned")) {
+      throw new IllegalStateException("Your account is banned from the forum.");
+    }
+    if (1 != (Integer) jsonObject.get("email:confirmed")) {
+      throw new IllegalStateException("Your email isn't confirmed yet!");
+    }
+  }
+
+  private Map<?, ?> queryUserInfo(final CloseableHttpClient client, final String username)
+      throws IOException {
+    final String encodedUsername = URLEncoder.encode(username, StandardCharsets.UTF_8);
```

Comment:
Note this fix for #6737 I added

Reply:
@RoiEXLab we should probably put the bug fix to its own PR:
- when this gets squashed, a bug fix being mixed in with a feature update will be somewhat confusing
- the bugfix needs to go out with the next release, which could be as soon as tonight or tomorrow.

Reply:
Would be nice to see a test case for this as well considering it was a bug. Tests are good for preventing future regressions, ensuring bugs are fixed well, and for giving feedback if the code is designed well, which should drive generally better code design (or at least prevent completely ad-hoc and  chaotic design).